### PR TITLE
dev/core#655 Don't sum on the contribution total_amount in contribution detail report

### PR DESCRIPTION
**Pull request by agh1**
Overview
----------------------------------------
The contribution amount field is a `SUM` in the contribution detail report but I can't think of any time you'd actually want that summed.  It causes Cartesian product problems when a contribution has multiple soft credits.

https://lab.civicrm.org/dev/core/issues/655

Before
----------------------------------------
Elina Adams has one $100 contribution with two soft credits: one for $100 and the other for $75.

Grouping by Contribution and Soft Credit shows the amount twice, next to each soft credit.
Screenshot_2019-03-27 Contribution Details grantdetailreport(1)

Grouping by Contribution (or selecting nothing in the Grouping tab): the amount is multiplied by the number of soft credits.
Screenshot_2019-03-27 Contribution Details grantdetailreport(3)

After
----------------------------------------
Again, Elina Adams has one $100 contribution with two soft credits: one for $100 and the other for $75.

Grouping by Contribution and Soft Credit shows the amount twice, next to each soft credit.
Screenshot_2019-03-27 Contribution Details grantdetailreport

Grouping by Contribution (or selecting nothing in the Grouping tab): the amount is still correct.
Screenshot_2019-03-27 Contribution Details grantdetailreport(2)

Technical Details
----------------------------------------
That field has been a sum [since the dawn of time (or Git)](https://github.com/civicrm/civicrm-core/blob/6a4880350680e1e4d20e5c8a622a791f926ca750/CRM/Report/Form/Contribute/Detail.php#L195).  I have no idea why.

### CODE REVIEW

File: CRM/Report/Form/Contribute/Detail.php [hunk 1]
```diff
@@ -160,7 +160,6 @@ public function __construct() {
           'total_amount' => array(
             'title' => ts('Amount'),
             'required' => TRUE,
-            'statistics' => array('sum' => ts('Amount')),
```
Comment (twomice) [inline CRM/Report/Form/Contribute/Detail.php]: @agh1 I agree about Cartesian product issues; that's my current beef with sum() in this usage. **But**, I think this line is what provides the "total contributions" statistic at the bottom of the report, and that a lot of people (who? "a lot of people" is all I know) rely on that bit of information in this report. Thoughts?

Comment (elisseck) [reply CRM/Report/Form/Contribute/Detail.php]: Hey Allen,

I'm reviewing this as well by request. Will post below shortly with a couple of screenshots.

**Comment by elisseck** [comment]
I tested and found the same behavior for CiviCRM 5.13.alpha1. Below is an image of the contribution detail report, showing both contributions and soft-credits. admin@example.com has a contribution for the amount of $25 and two soft credits to two different individuals, one for $25 and one for $10:

image

Post-patch I see the following report instead:

image

@agh1 @twomice my notes for this change are the following:

- Post-PR, the in-row amount appears correct. The summary statistics amount at the bottom still reads an incorrect $50 at "Total Amount (Contributions)", so we've made progress but that total amount at the bottom needs work as well. Soft-credit total amount appears correct.
- I inspected the failing tests, it looks like it's expecting this to go through a money util but instead it's just getting a number i.e. 
image


Other notes about the report are:

- Links for the soft-credit row are broken both pre and post-PR
- Total contributions statistic at the bottom lists "2" even if we filter for "Contributions Only" both pre and post-PR:

image

My suggestion is this is acceptable for an in-row fix if we can get it to pass the test, and we should open another small PR for the summary statistics themselves.

**Comment by agh1** [comment]
@twomice I think you could do statistics at the bottom without a field
being a statistic.  That's the `statistics()` function.  With these changes
it still has the totals and they appear to be accurate:
http://core-13906-8dkv1.test-1.civicrm.org:8001/civicrm/report/instance/8?force=1&reset=1

**Comment by JoeMurray** [comment]
There was a group of fundraisers who came to an early sprint that insisted this is what they wanted. It didn't make sense to me, but they were the domain experts.

**Comment by eileenmcnaughton** [comment]
I use that report WITHOUT soft credits and the sum thing IS useful then.

**Comment by twomice** [comment]
@elisseck has offered a good review and a good idea, namely:
- This fixes the amount in the row, and doesn't change behavior of end-of-report statistics.
- Thus it's probably good to get it to pass existing tests (and ideally write a new one for this fix) and merge this PR; then let a future PR address fixes to statistics.

**Comment by agh1** [comment]
I just added a commit that makes the test pass locally by formatting the amount in-row with the currency of the contribution.  I think the fact that the field was a sum triggered some handy-dandy formatter, and making it just a regular field bypassed that.

@eileenmcnaughton just to clarify this doesn't ditch the statistics at the bottom--as @elisseck points out, they were wrong and continue to be wrong if there are multiple soft credits on a contribution.  I think the only way to fix that would be to rework the joins on the statistics query (which may require the big rework that was discussed in dev/core#655).

**Comment by agh1** [comment]
Okay I think I found the more fundamental issue.  The function `joinContributionToSoftCredit()` is supposed to join the soft credits table only if we're displaying soft credits or both contributions and soft credits.  The condition there was whether `contribution_or_soft_value` was set at all--and it's always set.  I just changed it so that if `contribution_or_soft_value` is set to `contributions_only` (and no soft credit fields are displayed/filtered), it doesn't add the join.

The result is that so long as you ignore the soft credits, it should give accurate values and totals in the statistics.

**Comment by elisseck** [comment]
@agh1 I tested this latest push on the same instance and I now see the summary statistics working when the filter is set to 'contributions only'. The following images have the one confounding contribution with soft credits (see extra row in before etc), and the rest are just regular contributions no soft credit. Since group-by was mentioned above... I also did a group-by contact just to show the summation at the top of the group is also correct after this change:

Before:
image

After:
image

**Comment by MegaphoneJon** [comment]
IIRC the issue here doesn't just manifest when soft credits are present, but when multiple (partial) payments are made against a single contribution.

**Comment by agh1** [comment]
I just added two new commits, but I think this is a good stopping point.  One is a test that covers multiple soft credits.  The other is form validation so you don't attempt to group by soft credit when you're getting contributions only.  (Now that we don't join the soft credit table, you get a database error instead of just nonsense.)

@MegaphoneJon I couldn't recreate a problem with multiple payments on master, but testing with these changes appears to handle multiple payments accurately.

**Comment by eileenmcnaughton** [comment]
I take back my comments - this report only groups by contribution or soft credit - not other fields

**Comment by eileenmcnaughton** [comment]
OK so this seems to have 4 people involved - 2 are AGH people so if the other 2 of you (@twomice & @MegaphoneJon ) confirm you are happy with it then ping me to merge. I think your discussion / analysis is robust enough here!

**Comment by eileenmcnaughton** [comment]
@twomice  @MegaphoneJon did you reach a conclusion (also @lcdservices may have an interest)

**Comment by twomice** [comment]
I'm not able to perform a full review, but based on @elisseck 's review and other discussion above, I'm rescinding any concerns I've voiced so far. I.e., I'll abstain from voting on this motion.

**Comment by eileenmcnaughton** [comment]
OK I think I'm going to merge this based on a re-read @MegaphoneJon pipe up if you think this is wrong - we have a month of rc to re-dress

**Pull request merged**